### PR TITLE
Kernel: Remove a bunch of allocations

### PR DESCRIPTION
**Pull request by gunnarbeutner**
**AK: Don't unlink intrusive list elements in the destructor**

Removing the element from the intrusive linked list might not be safe if doing so requires a lock. Instead this is something the caller should have done so let's verify instead that we're not on any lists.

**Kernel: Remove an allocation when blocking a thread**

When blocking a thread with a timeout we would previously allocate a `Timer` object. This removes the allocation for that `Timer` object.

**Kernel: Use the Function class for deferred_call_queue()**

This avoids allocations for `deferred_call_queue()`.

**Kernel: Use the Function class for smp_broadcast()/smp_unicast()**

This avoids allocations for `smp_broadcast()` and `smp_unicast()` by using the `Function` class.

Allocations before:

Screenshot from 2021-05-19 23-09-44

And afterwards:

Screenshot from 2021-05-20 02-10-13

### CODE REVIEW

**Comment by gunnarbeutner** [comment]
Note to self:
```
[colonel(0:0)]: ASSERTION FAILED: false
[colonel(0:0)]: ../.././AK/Function.h:180 in AK::Function<Out(In ...)>::CallableWrapperBase* AK::Function<Out(In ...)>::callable_wrapper() const [with Out = void; In = {}]
[#0 colonel(0:0)]: 0xc05d3478  abort +0x4b
[#0 colonel(0:0)]: 0xc05c5edb  __assertion_failed(char const*, char const*, unsigned int, char const*) +0xd5
[#0 colonel(0:0)]: 0xc01457b0  AK::Function<void ()>::clear() +0x17e
[#0 colonel(0:0)]: 0xc05ced5b  Kernel::Processor::deferred_call_queue(AK::Function<void ()>) +0x18d
[#0 colonel(0:0)]: 0xc05436bb  Kernel::TimerQueue::fire()::{lambda(Kernel::TimerQueue::Queue&)#1}::operator()(Kernel::TimerQueue::Queue&) const +0xd3f
[#0 colonel(0:0)]: 0xc0543a46  Kernel::TimerQueue::fire() +0x108
[#0 colonel(0:0)]: 0xc053968a  Kernel::TimeManagement::system_timer_tick(Kernel::RegisterState const&) +0x104
[#0 colonel(0:0)]: 0xc0537646  Kernel::HardwareTimer<Kernel::IRQHandler>::handle_irq(Kernel::RegisterState const&) +0x272
[#0 colonel(0:0)]: 0xc0537065  Kernel::HPETComparator::handle_irq(Kernel::RegisterState const&) +0x5f
[#0 colonel(0:0)]: 0xc05d2951  handle_interrupt +0x6fd
[#0 colonel(0:0)]: 0xc05d43f9  interrupt_common_asm_entry +0x2b
[#0 colonel(0:0)]: 0xc05c4915  exit_kernel_thread +0x0
```

**Comment by gunnarbeutner** [comment]
Fixed, but wow, those explicit ctor/dtor calls are ugly.

File: AK/IntrusiveList.h [hunk 1]
```diff
@@ -320,8 +320,7 @@ namespace Detail {
 template<typename T, typename Container>
 inline IntrusiveListNode<T, Container>::~IntrusiveListNode()
 {
-    if (m_storage)
-        remove();
+    VERIFY(m_storage == nullptr);
```
Comment (bgianfo) [inline AK/IntrusiveList.h:323]: Nitpick: Should we just use `IntrusiveListNode::is_in_list()`?
```suggestion
    VERIFY(!is_in_list());
```

File: Kernel/Arch/x86/CPU.h [hunk 2]
```diff
@@ -608,19 +614,21 @@ struct ProcessorMessageEntry {
...
-            void* data;
-            void (*handler)(void*);
-            void (*free)(void*);
-        } callback_with_data;
-    };
-    bool have_data;
+    alignas(HandlerFunction) u8 handler_storage[sizeof(HandlerFunction)];
     bool was_allocated;
+
+    Function<void()>& handler_value()
```
Comment (bgianfo) [inline Kernel/Arch/x86/CPU.h:623]: Can't we use the using alias here as well? 
```suggestion
    HandlerFunction& handler_value()
```

**Review by bgianfo — APPROVED** [review]

File: Kernel/Arch/x86/CPU.h (hunk 2 above)
Comment (gunnarbeutner) [reply Kernel/Arch/x86/CPU.h]: Sure.

File: AK/IntrusiveList.h (hunk 1 above)
Comment (gunnarbeutner) [reply AK/IntrusiveList.h]: Fixed.

**Pull request merged**